man/default_store.3
.TH DEFAULT_STORE 3 "25 Jun 2002" V5.1.3.pre1 "Net-SNMP"
.UC 5
.SH NAME
default_store \- generic storage of global data.
.SH SYNOPSIS
.B #include <net-snmp/net-snmp-config.h>
.br
.B #include <net-snmp/config_api.h>

.BI "int netsnmp_ds_set_boolean(int " store ", int " which ", int " val ");"
.br
.BI "int netsnmp_ds_get_boolean(int " store ", int " which ");"
.br
.BI "int netsnmp_ds_set_int(int " store ", int " which ", int " val ");"
.br
.BI "int netsnmp_ds_get_int(int " store ", int " which ");"
.br
.BI "int netsnmp_ds_set_string(int " store ", int " which ", "
.br
.BI "                          const char *" val ");"
.br
.BI "char *netsnmp_ds_get_string(int " store ", int " which ");"
.br
.BI "int netsnmp_ds_register_config(u_char " type ", "
.br
.BI "                               const char *" ftype ","
.br
.BI "                               const char *" token ","
.br
.BI "                               int " store ", int " which ");"
.br
.BI "int netsnmp_ds_register_premib(u_char " type ", "
.br
.BI "                               const char *" ftype ","
.br
.BI "                               const char *" token ","
.br
.BI "                               int " store ", int " which ");"
.br
.BI "void netsnmp_ds_shutdown(void);"
.fi
.SH DESCRIPTION
The purpose of the default storage is three-fold:
.IP 1)
To create a global storage space without creating a whole bunch of
globally accessible variables or a whole bunch of access functions to
work with more privately restricted variables.
.IP 2)
To provide a single location where the thread locking needs to be
implemented. At the time of this writing, however, thread locking is not
yet in place.
.IP 3)
To reduce the number of cross dependencies between code pieces that
may or may not be linked together in the long run. This provides for a
single location in which configuration data, for example, can be
stored for a separate section of code that may not be linked in to
the application in question.
.PP
The functions defined here implement these goals.
.PP
Currently, three data types are supported: booleans, integers, and
strings. Each of these data types have separate storage
spaces. In addition, the storage space for each data type is divided
further by the application level.   Currently, there are two storage
spaces. The first is reserved for the SNMP library itself. The second 
is intended for use in applications and is not modified or checked by
the library, and, therefore, this is the space usable by you.
.PP
You can think of these storage spaces as being 3 arrays, something
like bool_storage[storeid][which], int_storage[storeid][which], and
string_storage[storeid][which].  The data is then accessed through the 
functions defined below.  For data you wish to store, you should use a 
.I store
parameter of
.BR NETSNMP_DS_APPLICATION_ID .
.PP
The storage space used by the library (for which the 
.I store
parameter is
.BR NETSNMP_DS_LIBRARY_ID )
is defined in the default_store.h file, which currently contains the 
following defines: 
.PP
.nf
     */
/**
 * @def NETSNMP_DS_LIBRARY_ID These definitions correspond with the "storid" argument to the API 
 */
#define NETSNMP_DS_LIBRARY_ID     0
#define NETSNMP_DS_APPLICATION_ID 1
#define NETSNMP_DS_TOKEN_ID       2

    /*
     * These definitions correspond with the "which" argument to the API,
     * when the storeid argument is NETSNMP_DS_LIBRARY_ID 
     */
    /*
     * library booleans 
     */
#define NETSNMP_DS_LIB_MIB_ERRORS          0
#define NETSNMP_DS_LIB_SAVE_MIB_DESCRS     1
#define NETSNMP_DS_LIB_MIB_COMMENT_TERM    2
#define NETSNMP_DS_LIB_MIB_PARSE_LABEL     3
#define NETSNMP_DS_LIB_DUMP_PACKET         4
#define NETSNMP_DS_LIB_LOG_TIMESTAMP       5
#define NETSNMP_DS_LIB_DONT_READ_CONFIGS   6
#define NETSNMP_DS_LIB_MIB_REPLACE         7    /* replace objects from latest module */
#define NETSNMP_DS_LIB_PRINT_NUMERIC_ENUM  8    /* print only numeric enum values */
#define NETSNMP_DS_LIB_PRINT_NUMERIC_OIDS  9    /* print only numeric enum values */
#define NETSNMP_DS_LIB_DONT_BREAKDOWN_OIDS 10   /* dont print oid indexes specially */
#define NETSNMP_DS_LIB_ALARM_DONT_USE_SIG  11   /* don't use the alarm() signal */
#define NETSNMP_DS_LIB_PRINT_FULL_OID      12   /* print fully qualified oids */
#define NETSNMP_DS_LIB_QUICK_PRINT         13   /* print very brief output for parsing */
#define NETSNMP_DS_LIB_RANDOM_ACCESS	   14   /* random access to oid labels */
#define NETSNMP_DS_LIB_REGEX_ACCESS	   15   /* regex matching to oid labels */
#define NETSNMP_DS_LIB_DONT_CHECK_RANGE    16   /* don't check values for ranges on send */
#define NETSNMP_DS_LIB_NO_TOKEN_WARNINGS   17   /* no warn about unknown config tokens */
#define NETSNMP_DS_LIB_NUMERIC_TIMETICKS   18   /* print timeticks as a number */
#define NETSNMP_DS_LIB_ESCAPE_QUOTES       19   /* shell escape quote marks in oids */
#define NETSNMP_DS_LIB_REVERSE_ENCODE      20   /* encode packets from back to front */
#define NETSNMP_DS_LIB_PRINT_BARE_VALUE	   21   /* just print value (not OID = value) */
#define NETSNMP_DS_LIB_EXTENDED_INDEX	   22   /* print extended index format [x1][x2] */
#define NETSNMP_DS_LIB_PRINT_HEX_TEXT      23   /* print ASCII text along with hex strings */
#define NETSNMP_DS_LIB_PRINT_UCD_STYLE_OID 24   /* print OID's using the UCD-style prefix suppression */
#define NETSNMP_DS_LIB_READ_UCD_STYLE_OID  25   /* require top-level OIDs to be prefixed with a dot */
#define NETSNMP_DS_LIB_HAVE_READ_PREMIB_CONFIG 26       /* have the pre-mib parsing config tokens been processed */
#define NETSNMP_DS_LIB_HAVE_READ_CONFIG    27   /* have the config tokens been processed */
#define NETSNMP_DS_LIB_QUICKE_PRINT        28   
#define NETSNMP_DS_LIB_DONT_PRINT_UNITS    29 /* don't print UNITS suffix */
#define NETSNMP_DS_LIB_NO_DISPLAY_HINT     30 /* don't apply DISPLAY-HINTs */
#define NETSNMP_DS_LIB_16BIT_IDS           31   /* restrict requestIDs, etc to 16-bit values */

    /*
     * library integers 
     */
#define NETSNMP_DS_LIB_MIB_WARNINGS  0
#define NETSNMP_DS_LIB_SECLEVEL      1
#define NETSNMP_DS_LIB_SNMPVERSION   2
#define NETSNMP_DS_LIB_DEFAULT_PORT  3
#define NETSNMP_DS_LIB_OID_OUTPUT_FORMAT  4
#define NETSNMP_DS_LIB_PRINT_SUFFIX_ONLY  NETSNMP_DS_LIB_OID_OUTPUT_FORMAT
#define NETSNMP_DS_LIB_STRING_OUTPUT_FORMAT 5

    /*
     * special meanings for the default SNMP version slot (NETSNMP_DS_LIB_SNMPVERSION) 
     */
#define NETSNMP_DS_SNMP_VERSION_1    128        /* bogus */
#define NETSNMP_DS_SNMP_VERSION_2c   1  /* real */
#define NETSNMP_DS_SNMP_VERSION_3    3  /* real */


    /*
     * library strings 
     */
#define NETSNMP_DS_LIB_SECNAME           0
#define NETSNMP_DS_LIB_CONTEXT           1
#define NETSNMP_DS_LIB_PASSPHRASE        2
#define NETSNMP_DS_LIB_AUTHPASSPHRASE    3
#define NETSNMP_DS_LIB_PRIVPASSPHRASE    4
#define NETSNMP_DS_LIB_OPTIONALCONFIG    5
#define NETSNMP_DS_LIB_APPTYPE           6
#define NETSNMP_DS_LIB_COMMUNITY         7
#define NETSNMP_DS_LIB_PERSISTENT_DIR    8
#define NETSNMP_DS_LIB_CONFIGURATION_DIR 9
#define NETSNMP_DS_LIB_SECMODEL          10
#define NETSNMP_DS_LIB_MIBDIRS           11
#define NETSNMP_DS_LIB_OIDSUFFIX         12
#define NETSNMP_DS_LIB_OIDPREFIX         13
#define NETSNMP_DS_LIB_CLIENT_ADDR       14
#define NETSNMP_DS_LIB_TEMP_FILE_PATTERN 15
#define NETSNMP_DS_LIB_KSM_KEYTAB        16
#define NETSNMP_DS_LIB_KSM_SERVICE_NAME  17


    /*
.fi
.SH FUNCTIONS
.TP
.BI "int netsnmp_ds_set_boolean(int " store ", int " which ", int " val ");"
Stores
.I TRUE
if 
.I val
!= 0 or else
.I FALSE
into the bool_storage[store][which] slot.  Returns
.B SNMPERR_GENERR
if the
.IR store " and " which
parameters do not correspond to a valid slot, or
.B SNMPERR_SUCCESS
otherwise.
.TP
.BI "int netsnmp_ds_get_boolean(int " store ", int " which ");"
Returns 1 if bool_storage[store][which] is 
.IR TRUE
or 0 if not.  May also return
.B SNMPERR_GENERR
if the 
.IR store " and " which
parameters do not correspond to a valid slot.
.TP
.BI "int netsnmp_ds_set_string(int " store ", int " which ", const char *" val ");"
Stores
.I val
into the string_storage[store][which] slot.  Returns
.B SNMPERR_SUCCESS
normally, or 
.B SNMPERR_GENERR
if the 
.IR store " and " which
parameters do not correspond to a valid slot.
.TP
.BI "char *netsnmp_ds_get_string(int " store ", int " which ");"
Returns the string which has been stored in the
string_storage[store][which] slot, or
.B NULL
if the
.IR store " and " which
parameters do not correspond to a valid slot.
.TP 
.BI "netsnmp_ds_set_int(int " store ", int " which ", int " val ");"
Stores
.I val
into the int_storage[store][which] slot.  Returns
.B SNMPERR_GENERR
if the
.IR store " and " which
parameters do not correspond to a valid slot, or
.B SNMPERR_SUCCESS
otherwise.
.TP
.BI "int netsnmp_ds_get_int(int " store ", int " which ");"
Returns the integer which has been stored in the
int_storage[store][which] slot, or
.B SNMPERR_GENERR
if the
.IR store " and " which
parameters do not correspond to a valid slot.
.TP
.BI "void netsnmp_ds_shutdown(void);"
Reclaims memory used to hold information gathered by
.BR netsnmp_ds_register_config " and " netsnmp_ds_register_premib .
.TP
.BI "int netsnmp_ds_register_config(u_char " type ", const char *" ftype ", const char *" token ", int " store ", int " which ");"
Registers a configuration file directive 
.I token
and attaches it to 
a default storage type and slot.  Specifically,
.I store
and
.I which
indicate the storage slot in the data type indicated by
.I type,
where
.I type
is one of the following constants:
.BR ASN_BOOLEAN ", " ASN_INTEGER ", or " ASN_OCTET_STR .
The
.I ftype
variable indicates the file name base string searched for the
.I token
keyword.  For example, the following call:
.RS
.IP
netsnmp_ds_register_config(ASN_INTEGER, "snmp", "testtoken", DS_APPLICATION_ID, 5)
.RE
.IP
would indicate that when the snmp.conf file(s) were found and parsed,
that any line beginning with the word "testtoken" should be read and
the value after "testtoken" should be stored into the
int_storage[DS_APPLICATION_ID][5] slot.  For example the following
line in the configuration file:
.RS
.IP
testtoken 502
.RE
.IP
would set int_storage[DS_APPLICATION_ID][5] = 502.  This function returns
.B SNMPERR_SUCCESS
if the registration was made successfully, or 
.B SNMPERR_GENERR
if the registration was not made (perhaps because the
.IR store " and " which
parameters do not correspond to a valid slot, or because of a memory
allocation failure).
.TP
.BI "int netsnmp_ds_register_premib(u_char " type ", const char *" ftype ", const char *" token ", int " store ", int " which ");"
Analogous to the preceeding function, but the
.I token
is processed before MIBs are read (this is therefore useful for controlling
MIB processing options).
.SH "SEE ALSO"
.BR snmp_config "(5), " read_config "(3)"



NEWS
This file contains major changes in released revisions.  Please read
the ChangeLog file for a more detailed listing of changes/bug fixes.

*5.0.10.2*
 Security:
   - Fixed a denial of service vulnerability when stream sockets have
     been configured for use (E.G., TCP but not UDP).

*5.0.10*
 Security:
   - If you use the fixproc script, you should immediately update to
     the version in this release to fix a temporary file race condition.

 Fixes:
   - Many.  This is a final release of the 5.0.X line of the Net-SNMP
     code base and we've accumulated a large number of fixes in this
     branch.  5.0.10 will be the last release from this line.
   - get rid of strtok (patch 1040330, backported by Thomas Anders, fixes 
     bug 1040686)

 Ports:

   Win32:
     - Cygwin compiler fixes

   Sun:
     - Many improvements and fixes supplied by Sun as they changed
       their SNMP agent from their's to ours for their Solaris X release.


*5.0.9*

 SECURITY:
    - An existing user/community could get access to data in MIB
      objects that were explicitly excluded from their view.

 Fixes:
    - Perl build environment should better under Windows
    - Misc kerberos support fixes.
    - Improvements on various manual pages.
    - A annoying bug with SETs being passed to pass scripts was fixed.
    - The often talked about VACM optimization improvement was fixed again.
    - mib2c handles augmentation tables better now.
    - Various 64 bit issues have been addressed.

*5.0.8*
 New:
    - No new features will be added to the 5.0.x line.

 Ports:
    - Update libtool to version 1.4.3, for the benefit of Darwin
    - diskio support for Darwin
    - Updates for OpenBSD 3. 
    - Updates to solaris README

 Fixes:
    - find libwrap w/nsl on RedHat
    - fix for openssl 0.9.7
    - Fix some AgentX memory leaks
    - use macro for inline function prototypes
    - Attempt to find unused port before running tests
    - Use SNMP_SLEEP environment variable when running tests
    - calculate a proper ifSpeed under linux when possible
    - better daemonization of snmpd
    - close and reopen snmptrapd log files on HUP
    - support for 16 bit reuqest ids
    - Recognize new 't' code in display hints
    - misc other fixes

*5.0.7*
 New:
    - VACM (access control) optimizations which will greatly benefit
      people who wish to exclude large portions of the MIB tree from
      some people.  Previously this was a large resource drain.
    - Add command line option to snmpd to set syslog facility
    - Reverse DISPLAY-HINT processing, i.e. it allows you to input data
      formatted like a DISPLAY-HINT prescribes
    - Support setting of sysDescr and sysObjectID via snmpd.conf
      configuration directives
    - New output option to force display of strings as hex
    - Persistent directory can be specified at runtime
    - Add support for Linux virtual interfaces in the ipAddressTable.
    - implemented the mteEventTable and the mteEventNotificationTable
      form the DISMAN-EVENT-MIB.

 Fixes:
    - AgentX no longer flagged as experimental
    - A few memory leak fixes for the table_iterator agent API.
    - Processed flag cleared before each pass of a set request
    - Remove snmpd pid file on exit
    - Restore default behaviour of building shared libraries
    - misc other fixes

*5.0.6*

 Fixes:
    - prevent denial of service attack from authenticated users
      (really this time).
    - misc other fixes

*5.0.5*

 New:
    - Support for OpenSSL 0.9.7
    - Beginning of support for AES encryption.
      ( Currently only usable with Net-SNMP and OpenSNMP software. )

 Ports:
    - win32 new project files win32sdk.dsw for those with the win32
      platform sdk.
    - win32 builds should work properly again, minus callbacks and thus
      AgentX.

 Fixes:
    - Several AgentX (the subagent protocol) specific bugs have been fixed
    - prevent denial of service attack from authenticated users
    - many many misc fixes
    - The perl agent module plugin should now work.
    - Many other perl module related fixes.
    - net-snmp-config --ldflags properly supplies -L flags
    - misc make test fixes and improvements.

*5.0.4*

   During the testing for release 5.0.4, a release candidate was uploaded
   to Sourceforge and marked as hidden. For variouse reasons, the release
   was delayed and further changes and fixes were made.  However, the 
   release candidate files still appeared on the public FTP server and our
   mirrors. To avoid the confusion that could occur to users who may have
   downloaded this release candidate, we are bumping the revision number
   to 5.0.5.  If you downloaded net-snmp-5.0.4.tar.gz, please upgrade
   to net-snmp-5.0.5.tar.gz.

*5.0.3*

 New:
    - the "dist" directory contains the beginnings of some init
      scripts and packaging utilities.

 Ports:
    - win32 builds should work properly again.

 Fixes:
    - the 5.0.2 package had a few broken packaging problems.  Opps.
    - a few misc fixes.

 Perl:
    - the perl modules should work with older versions of perl again.

*5.0.2*

 New:
    - mib2c support for the old-4.X style api and for scalars.
    - many improvements to the netsh shell environment.

 Fixes:
    - tcpwrappers support should work significantly better.
    - NFS file systems are better supported in the hrStorageTable
    - the memory reporting no longer accidentally reports -1 integers
    - a few memory leaks fixed in the table_iterator API
    - fixes for Sun's "make"
    - ds_ prefix name-space protection (-> netsnmp_ds_)
      (run configure with --enable-ucd-compatibility for old API)
    - v1/v2c access control fixes for ipv6 addresses
    - many many misc fixes

 Perl:
    - The perl modules have a number of fixes, and the build process
      was fixed (again).
    - all make test suites should pass for people now.
      (there are still some known minor limitations, but we removed the tests)

*5.0.1*

 New:

perl/default_store/default_store.pm
package NetSNMP::default_store;

use strict;
use Carp;

require Exporter;
require DynaLoader;
use AutoLoader;

use vars qw(@ISA %EXPORT_TAGS @EXPORT_OK @EXPORT $VERSION $AUTOLOAD);

@ISA = qw(Exporter DynaLoader);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use NetSNMP::default_store ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
%EXPORT_TAGS = ( 'all' => [ qw(
				   NETSNMP_DS_MAX_IDS
				   NETSNMP_DS_MAX_SUBIDS
				   NETSNMP_DS_LIBRARY_ID
				   NETSNMP_DS_APPLICATION_ID
				   NETSNMP_DS_TOKEN_ID
				   NETSNMP_DS_LIB_MIB_ERRORS
				   NETSNMP_DS_LIB_SAVE_MIB_DESCRS
				   NETSNMP_DS_LIB_MIB_COMMENT_TERM
				   NETSNMP_DS_LIB_MIB_PARSE_LABEL
				   NETSNMP_DS_LIB_DUMP_PACKET
				   NETSNMP_DS_LIB_LOG_TIMESTAMP
				   NETSNMP_DS_LIB_DONT_READ_CONFIGS
				   NETSNMP_DS_LIB_MIB_REPLACE
				   NETSNMP_DS_LIB_PRINT_NUMERIC_ENUM
				   NETSNMP_DS_LIB_PRINT_NUMERIC_OIDS
				   NETSNMP_DS_LIB_DONT_BREAKDOWN_OIDS
				   NETSNMP_DS_LIB_ALARM_DONT_USE_SIG
				   NETSNMP_DS_LIB_PRINT_FULL_OID
				   NETSNMP_DS_LIB_QUICK_PRINT
				   NETSNMP_DS_LIB_RANDOM_ACCESS
				   NETSNMP_DS_LIB_REGEX_ACCESS
				   NETSNMP_DS_LIB_DONT_CHECK_RANGE
				   NETSNMP_DS_LIB_NO_TOKEN_WARNINGS
				   NETSNMP_DS_LIB_NUMERIC_TIMETICKS
				   NETSNMP_DS_LIB_ESCAPE_QUOTES
				   NETSNMP_DS_LIB_REVERSE_ENCODE
				   NETSNMP_DS_LIB_PRINT_BARE_VALUE
				   NETSNMP_DS_LIB_EXTENDED_INDEX
				   NETSNMP_DS_LIB_PRINT_HEX_TEXT
				   NETSNMP_DS_LIB_PRINT_UCD_STYLE_OID
				   NETSNMP_DS_LIB_READ_UCD_STYLE_OID
				   NETSNMP_DS_LIB_HAVE_READ_PREMIB_CONFIG
				   NETSNMP_DS_LIB_HAVE_READ_CONFIG
				   NETSNMP_DS_LIB_QUICKE_PRINT
				   NETSNMP_DS_LIB_MIB_WARNINGS
				   NETSNMP_DS_LIB_SECLEVEL
				   NETSNMP_DS_LIB_SNMPVERSION
				   NETSNMP_DS_LIB_DEFAULT_PORT
				   NETSNMP_DS_LIB_OID_OUTPUT_FORMAT
				   NETSNMP_DS_SNMP_VERSION_1
				   NETSNMP_DS_SNMP_VERSION_2c
				   NETSNMP_DS_SNMP_VERSION_3
				   NETSNMP_DS_LIB_SECNAME
				   NETSNMP_DS_LIB_CONTEXT
				   NETSNMP_DS_LIB_PASSPHRASE
				   NETSNMP_DS_LIB_AUTHPASSPHRASE
				   NETSNMP_DS_LIB_PRIVPASSPHRASE
				   NETSNMP_DS_LIB_OPTIONALCONFIG
				   NETSNMP_DS_LIB_APPTYPE
				   NETSNMP_DS_LIB_COMMUNITY
				   NETSNMP_DS_LIB_PERSISTENT_DIR
				   NETSNMP_DS_LIB_CONFIGURATION_DIR
				   NETSNMP_DS_LIB_SECMODEL
				   NETSNMP_DS_LIB_MIBDIRS
	netsnmp_ds_get_boolean
	netsnmp_ds_get_int
	netsnmp_ds_get_string
	netsnmp_ds_get_void
	netsnmp_ds_register_config
	netsnmp_ds_register_premib
	netsnmp_ds_set_boolean
	netsnmp_ds_set_int
	netsnmp_ds_set_string
	netsnmp_ds_set_void
	netsnmp_ds_shutdown
	netsnmp_ds_toggle_boolean
) ] );

@EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

@EXPORT = qw(
				   NETSNMP_DS_MAX_IDS
				   NETSNMP_DS_MAX_SUBIDS
				   NETSNMP_DS_LIBRARY_ID
				   NETSNMP_DS_APPLICATION_ID
				   NETSNMP_DS_TOKEN_ID
				   NETSNMP_DS_LIB_MIB_ERRORS
				   NETSNMP_DS_LIB_SAVE_MIB_DESCRS
				   NETSNMP_DS_LIB_MIB_COMMENT_TERM
				   NETSNMP_DS_LIB_MIB_PARSE_LABEL
				   NETSNMP_DS_LIB_DUMP_PACKET
				   NETSNMP_DS_LIB_LOG_TIMESTAMP
				   NETSNMP_DS_LIB_DONT_READ_CONFIGS
				   NETSNMP_DS_LIB_MIB_REPLACE
				   NETSNMP_DS_LIB_PRINT_NUMERIC_ENUM
				   NETSNMP_DS_LIB_PRINT_NUMERIC_OIDS
				   NETSNMP_DS_LIB_DONT_BREAKDOWN_OIDS
				   NETSNMP_DS_LIB_ALARM_DONT_USE_SIG
				   NETSNMP_DS_LIB_PRINT_FULL_OID
				   NETSNMP_DS_LIB_QUICK_PRINT
				   NETSNMP_DS_LIB_RANDOM_ACCESS
				   NETSNMP_DS_LIB_REGEX_ACCESS
				   NETSNMP_DS_LIB_DONT_CHECK_RANGE
				   NETSNMP_DS_LIB_NO_TOKEN_WARNINGS
				   NETSNMP_DS_LIB_NUMERIC_TIMETICKS
				   NETSNMP_DS_LIB_ESCAPE_QUOTES
				   NETSNMP_DS_LIB_REVERSE_ENCODE
				   NETSNMP_DS_LIB_PRINT_BARE_VALUE
				   NETSNMP_DS_LIB_EXTENDED_INDEX
				   NETSNMP_DS_LIB_PRINT_HEX_TEXT
				   NETSNMP_DS_LIB_PRINT_UCD_STYLE_OID
				   NETSNMP_DS_LIB_READ_UCD_STYLE_OID
				   NETSNMP_DS_LIB_HAVE_READ_PREMIB_CONFIG
				   NETSNMP_DS_LIB_HAVE_READ_CONFIG
				   NETSNMP_DS_LIB_QUICKE_PRINT
				   NETSNMP_DS_LIB_MIB_WARNINGS
				   NETSNMP_DS_LIB_SECLEVEL
				   NETSNMP_DS_LIB_SNMPVERSION
				   NETSNMP_DS_LIB_DEFAULT_PORT
				   NETSNMP_DS_LIB_OID_OUTPUT_FORMAT
				   NETSNMP_DS_SNMP_VERSION_1
				   NETSNMP_DS_SNMP_VERSION_2c
				   NETSNMP_DS_SNMP_VERSION_3
				   NETSNMP_DS_LIB_SECNAME
				   NETSNMP_DS_LIB_CONTEXT
				   NETSNMP_DS_LIB_PASSPHRASE
				   NETSNMP_DS_LIB_AUTHPASSPHRASE
				   NETSNMP_DS_LIB_PRIVPASSPHRASE
				   NETSNMP_DS_LIB_OPTIONALCONFIG
				   NETSNMP_DS_LIB_APPTYPE
				   NETSNMP_DS_LIB_COMMUNITY
				   NETSNMP_DS_LIB_PERSISTENT_DIR
				   NETSNMP_DS_LIB_CONFIGURATION_DIR
				   NETSNMP_DS_LIB_SECMODEL
				   NETSNMP_DS_LIB_MIBDIRS
);
$VERSION = '5.2.1.2';

sub AUTOLOAD {
    # This AUTOLOAD is used to 'autoload' constants from the constant()
    # XS function.  If a constant is not found then control is passed
    # to the AUTOLOAD in AutoLoader.

    my $constname;
    ($constname = $AUTOLOAD) =~ s/.*:://;
    croak "& not defined" if $constname eq 'constant';
    my $val = constant($constname, @_ ? $_[0] : 0);
    if ($! != 0) {
	if ($! =~ /Invalid/ || $!{EINVAL}) {
	    $AutoLoader::AUTOLOAD = $AUTOLOAD;
	    goto &AutoLoader::AUTOLOAD;
	}
	else {
	    croak "Your vendor has not defined NetSNMP::default_store macro $constname";
	}
    }
    {
	no strict 'refs';
	# Fixed between 5.005_53 and 5.005_61
#	if ($] >= 5.00561) {
#	    *$AUTOLOAD = sub () { $val };
#	}
#	else {
	    *$AUTOLOAD = sub { $val };
#	}
    }
    goto &$AUTOLOAD;
}

bootstrap NetSNMP::default_store $VERSION;

# Preloaded methods go here.

# Autoload methods go after =cut, and are processed by the autosplit program.

1;
__END__

=head1 NAME

NetSNMP::default_store - Perl extension for blah blah blah

=head1 SYNOPSIS

  use NetSNMP::default_store;
  $port = netsnmp_ds_get_int(NETSNMP_DS_LIBRARY_ID, DS_LIB_DEFAULT_PORT);
  netsnmp_ds_set_int(NETSNMP_DS_LIBRARY_ID, DS_LIB_DEFAULT_PORT, 161);

=head1 DESCRIPTION

This module is a wrapper around the net-snmp default store routines.
See the net-snmp default_store manual page for details on what the
various functions do and the values that can be set/retrieved.

=head2 EXPORT

None by default.

=head2 Exportable constants

				   NETSNMP_DS_MAX_IDS
				   NETSNMP_DS_MAX_SUBIDS
				   NETSNMP_DS_LIBRARY_ID
				   NETSNMP_DS_APPLICATION_ID
				   NETSNMP_DS_TOKEN_ID
				   NETSNMP_DS_LIB_MIB_ERRORS
				   NETSNMP_DS_LIB_SAVE_MIB_DESCRS
				   NETSNMP_DS_LIB_MIB_COMMENT_TERM
				   NETSNMP_DS_LIB_MIB_PARSE_LABEL
				   NETSNMP_DS_LIB_DUMP_PACKET
				   NETSNMP_DS_LIB_LOG_TIMESTAMP
				   NETSNMP_DS_LIB_DONT_READ_CONFIGS
				   NETSNMP_DS_LIB_MIB_REPLACE
				   NETSNMP_DS_LIB_PRINT_NUMERIC_ENUM
				   NETSNMP_DS_LIB_PRINT_NUMERIC_OIDS
				   NETSNMP_DS_LIB_DONT_BREAKDOWN_OIDS
				   NETSNMP_DS_LIB_ALARM_DONT_USE_SIG
				   NETSNMP_DS_LIB_PRINT_FULL_OID
				   NETSNMP_DS_LIB_QUICK_PRINT
				   NETSNMP_DS_LIB_RANDOM_ACCESS
				   NETSNMP_DS_LIB_REGEX_ACCESS
				   NETSNMP_DS_LIB_DONT_CHECK_RANGE
				   NETSNMP_DS_LIB_NO_TOKEN_WARNINGS
				   NETSNMP_DS_LIB_NUMERIC_TIMETICKS
				   NETSNMP_DS_LIB_ESCAPE_QUOTES
				   NETSNMP_DS_LIB_REVERSE_ENCODE
				   NETSNMP_DS_LIB_PRINT_BARE_VALUE
				   NETSNMP_DS_LIB_EXTENDED_INDEX
				   NETSNMP_DS_LIB_PRINT_HEX_TEXT
				   NETSNMP_DS_LIB_PRINT_UCD_STYLE_OID
				   NETSNMP_DS_LIB_READ_UCD_STYLE_OID
				   NETSNMP_DS_LIB_HAVE_READ_PREMIB_CONFIG
				   NETSNMP_DS_LIB_HAVE_READ_CONFIG
				   NETSNMP_DS_LIB_QUICKE_PRINT
				   NETSNMP_DS_LIB_MIB_WARNINGS
				   NETSNMP_DS_LIB_SECLEVEL
				   NETSNMP_DS_LIB_SNMPVERSION
				   NETSNMP_DS_LIB_DEFAULT_PORT
				   NETSNMP_DS_LIB_OID_OUTPUT_FORMAT
				   NETSNMP_DS_SNMP_VERSION_1
				   NETSNMP_DS_SNMP_VERSION_2c
				   NETSNMP_DS_SNMP_VERSION_3
				   NETSNMP_DS_LIB_SECNAME
				   NETSNMP_DS_LIB_CONTEXT
				   NETSNMP_DS_LIB_PASSPHRASE
				   NETSNMP_DS_LIB_AUTHPASSPHRASE
				   NETSNMP_DS_LIB_PRIVPASSPHRASE
				   NETSNMP_DS_LIB_OPTIONALCONFIG
				   NETSNMP_DS_LIB_APPTYPE
				   NETSNMP_DS_LIB_COMMUNITY
				   NETSNMP_DS_LIB_PERSISTENT_DIR
				   NETSNMP_DS_LIB_CONFIGURATION_DIR
				   NETSNMP_DS_LIB_SECMODEL
				   NETSNMP_DS_LIB_MIBDIRS

=head2 Exportable functions

  int netsnmp_ds_get_boolean(int storeid, int which)
  int netsnmp_ds_get_int(int storeid, int which)
  char *netsnmp_ds_get_string(int storeid, int which)
  void *netsnmp_ds_get_void(int storeid, int which)
  int netsnmp_ds_register_config(unsigned char type, const char *ftype, const char *token,
                       int storeid, int which)
  int netsnmp_ds_register_premib(unsigned char type, const char *ftype, const char *token,
                       int storeid, int which)
  int netsnmp_ds_set_boolean(int storeid, int which, int value)
  int netsnmp_ds_set_int(int storeid, int which, int value)
  int netsnmp_ds_set_string(int storeid, int which, const char *value)
  int netsnmp_ds_set_void(int storeid, int which, void *value)
  void netsnmp_ds_shutdown(void)
  int netsnmp_ds_toggle_boolean(int storeid, int which)

=head1 AUTHOR

Wes Hardaker, hardaker@users.sourceforge.net

=head1 SEE ALSO

perl(1), default_store(3).

=cut
